src/librustc_mir_build/build/matches/mod.rs

```diff
@@ -308,25 +302,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 guard,
                 fake_borrow_temps,
                 scrutinee_span,
+                true,
             )
         } else {
             let target_block = self.cfg.start_new_block();
-
+            let mut schedule_drops = true;
             // We keep a stack of all of the bindings and type asciptions
             // from the the parent candidates that we visit, that also need to
             // be bound for each candidate.
             traverse_candidate(
                 candidate,
                 &mut Vec::new(),
                 &mut |leaf_candidate, parent_bindings| {
-                    self.clear_top_scope(arm_scope);
+                    if let Some(arm_scope) = arm_scope {
+                        // Avoid scheduling drops multiple times by unscheduling drops.
+                        self.clear_top_scope(arm_scope);
+                    }
                     let binding_end = self.bind_and_guard_matched_candidate(
                         leaf_candidate,
                         parent_bindings,
                         guard,
                         &fake_borrow_temps,
                         scrutinee_span,
+                        schedule_drops,
                     );
+                    if arm_scope.is_none() {
+                        // If we aren't in a match, then our bindings may not be
+                        // the only thing in the top scope, so only schedule
+                        // them to drop for the first pattern instead.
```

Comment:
I don't understand the argument presented here. 

But maybe part of my problem is that I cannot tell if this is being done as an efficiency hack, or as something to ensure correctness of the generated code.

Can you spell it out a bit more? I don't know if I need an example or what.